### PR TITLE
Add support for CudaHalfTensor, CudaDoubleTensor training

### PR DESCRIPTION
**Pull request by gchanan**
via the -precision parameter.

Note that this does not add support for the pre-trained models
in classify, extract-features.

### CODE REVIEW

File: train.lua [hunk 1]
```diff
@@ -154,9 +154,9 @@ function Trainer:copyInputs(sample)
    -- Copies the input to a CUDA tensor, if using 1 GPU, or to pinned memory,
    -- if using DataParallelTable. The target is always copied to a CUDA tensor
    self.input = self.input or (self.opt.nGPU == 1
-      and torch.CudaTensor()
+      and torch[self.opt.tensorType:match('torch.(%a+)')]()
       or cutorch.createCudaHostTensor())
-   self.target = self.target or (torch.CudaLongTensor and torch.CudaLongTensor()or torch.CudaTensor())
+   self.target = self.target or (torch.CudaLongTensor and torch.CudaLongTensor() or torch[self.opt.tensorType:match('torch.(%a+)')]())
```
Comment (colesbury) [inline train.lua]: Shouldn't this always be `torch.CudaLongTensor`?

Comment (gchanan) [inline train.lua]: Yes.

**Comment by gchanan** [comment]
note: this currently works for 1 gpu, may need more work for greater number of GPUs.

**Comment by ngimel** [comment]
@gchanan, yes, for multiGPU it would need more work. DataParallelTable wants all the parameters to be of the same type, and batchnorm (grrrr) has float parameters even when input/output are fp16. What we've discovered with @szagoruyko's wide resnets, if batchnorm parameters are converted to fp16, accuracy slightly suffers.

**Comment by gchanan** [comment]
@ngimel thanks!  I meant the above as a note that we shouldn't merge this without at a minimum, some additional configuration checking.  I'll look at what we'd need to do in DataParallelTable to support this.

**Comment by gchanan** [comment]
that's right, this probably needs modified versions of other libraries as well (cutorch, cunn via DataParallelTable).  I'll close that until those are in better shape.
